src/common/flux/async-action.ts

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { Action } from 'common/flux/action';
+import { ScopeMutex } from 'common/flux/scope-mutex';
+
+export class AsyncAction<TPayload> implements Action<TPayload, Promise<void>> {
+    private listeners: ((payload: TPayload) => Promise<void>)[] = [];
+
+    constructor(private readonly scopeMutex: ScopeMutex = new ScopeMutex()) {}
+
+    public async invoke(payload: TPayload, scope?: string): Promise<void> {
+        this.scopeMutex.tryLockScope(scope);
+
+        try {
+            const results = await Promise.allSettled(
+                this.listeners.map((listener: (payload: TPayload) => void) => listener(payload)),
+            );
+            this.throwCombinedRejections(results);
+        } finally {
+            this.scopeMutex.unlockScope(scope);
+        }
+    }
+
+    public addListener(listener: (payload: TPayload) => Promise<void>): void {
+        this.listeners.push(listener);
+    }
+
+    private throwCombinedRejections(promiseResults: PromiseSettledResult<void>[]): void {
```

Comment:
My understanding was that we would combine listener results with the `EventResposeFactory`. This seems to be duplicating a lot of the logic in `mergeAsyncResponses` anyways, can we call that here instead?

Reply:
(Discussed offline with Dan) EventResponseFactory is a bit more involved than what we need here: it also adds a delay to fire-and-forget handlers, which we wouldn't need to do here, and takes some additional dependencies that we wouldn't need to use. I refactored EventResponseFactory to extract only the logic of mergeAsyncResponses so we can reuse that here.